### ISSUE TITLE
"Hide most of non-public..." PR breaks Windows build

### ISSUE THREAD

**Issue by noelotterness** [jbeder/yaml-cpp#1046]
PR #984 broke the Windows use of the yaml-cpp branch. It was fixed when it was backed out on Sept 28th and it has broken the Windows usage again with the revert of the revert.
Here is what happens with the broken submission:
```"D:\a\IIS-Module\IIS-Module\main\Native-IIS.vcxproj" (default target) (2) ->
(ClCompile target) -> 
  D:\a\IIS-Module\IIS-Module\yaml-cpp\include\yaml-cpp\parser.h(13,10): fatal error C1083: Cannot open include file: 'yaml-cpp/dll.h': No such file or directory [D:\a\IIS-Module\IIS-Module\main\Native-IIS.vcxproj]
  D:\a\IIS-Module\IIS-Module\yaml-cpp\include\yaml-cpp\parser.h(13,10): fatal error C1083: Cannot open include file: 'yaml-cpp/dll.h': No such file or directory [D:\a\IIS-Module\IIS-Module\main\Native-IIS.vcxproj]
  D:\a\IIS-Module\IIS-Module\yaml-cpp\include\yaml-cpp\parser.h(13,10): fatal error C1083: Cannot open include file: 'yaml-cpp/dll.h': No such file or directory [D:\a\IIS-Module\IIS-Module\main\Native-IIS.vcxproj]
  D:\a\IIS-Module\IIS-Module\yaml-cpp\include\yaml-cpp\parser.h(13,10): fatal error C1083: Cannot open include file: 'yaml-cpp/dll.h': No such file or directory [D:\a\IIS-Module\IIS-Module\main\Native-IIS.vcxproj]
```

**Comment by noelotterness** [comment]
First I cannot provide the source code since it is not a public repository. The build system files are what are generated by Visual Studio to build a native C++ DLL for IIS. I include the directories where the generated source is placed. The problem is that there is something in the yaml-cpp source that requires the dll.h file and that is no longer being generated for the Windows 2019 build.
Perhaps it has been moved to a new location?
The Project include directive is for "yaml-cpp\include" with some pathing to obtain the correct location based upon either a local build or a GitHub build.
The problem is that it all used to work prior to your commit and the code was configured as described in the documentation. If there is something different I need to do to make it work now (also I do not use ExternalProject_Add()), that would be great.

**Comment by pinotree** [comment]
> The build system files are what are generated by Visual Studio to build a native C++ DLL for IIS. I include the directories where the generated source is placed.

So you are embedding yaml-cpp are part of your sources?

> The problem is that there is something in the yaml-cpp source that requires the dll.h file and that is no longer being generated for the Windows 2019 build.

`dll.h` is still available during a build (no matter the OS). What changed is that, instead of a static file as `include/yaml-cpp/dll.h` relative to the yaml-cpp source directory, now it is generated by cmake and it is available as `include/yaml-cpp/dll.h` relative to the yaml-cpp **build** directory. So, in case you are manually adding include directories to use the embedded yaml-cpp copy, you have to add one more include path; again, see https://github.com/jbeder/yaml-cpp/issues/1036#issuecomment-929475353.

**Comment by PhilipDeegan** [comment]
> `dll.h` is still available during a build (no matter the OS). What changed is that, instead of a static file as `include/yaml-cpp/dll.h` relative to the yaml-cpp source directory, now it is generated by cmake 

what about anyone not using cmake?

**Comment by pinotree** [comment]
> what about anyone not using cmake?

Considering the build system is cmake, it seems fair to assume that the build system will take care of everything needed to properly build yaml-cpp.

If you want to embed yaml-cpp and build it yourself without using cmake, then you will need to replicate **all the logic** yourself:
- keep up with the list of sources (which can be added and removed any time)
- build the needed libraries/tools
- make sure the right compiler flags are used
- properly discover build/runtime dependencies
- create helper config.h-alike headers with various bits of information (usually about external dependencies)

This was valid also before, and usually it is valid in any project for any build system different than one one(s) supported by the project.

**Comment by hkarel** [comment]
> @pinotree, is there a way to solve the issue you were trying to solve (if I recall, reducing the non-public symbols that are exported) without generating the dll.h file?

@pinotree, I would also like to hear the answer to this question. 
I think that it was possible to completely do without the dynamic generation of the dll.h file. To do this, it was necessary to figure out how to use the dllexport/dllimport directives for the MS compiler. Why @pinotree didn't want to do this - question! 

> ... yaml-cpp is so simple that you can usually just throw all of the sources into your build system and it'll just work.

It really is! In this I see a big plus of yaml-cpp. In my projects, I include yaml-cpp in the form of source codes, the assembly of the library takes place in conjunction with a specific project.  CMake is not used at all. Therefore, after updating to commit 0733aeb451e2526dc05824d954348ca5baa0feed, all my projects stopped building. Of course the situation will be corrected - I will return back the static file dll.h. 

In fact, now cmake has received a monopoly to build the yaml-cpp. Those who do not use cmake in their work will have to find solutions on their own. It seems to me that this is not very correct.

@jbeder, did you manage to build the master-branch using Bazel?

**Comment by hkarel** [comment]
The updated dll.h file may look like [this](https://github.com/hkarel/yaml-cpp/blob/b0771c76840a1f18303d11539441b7fe853b66cf/include/yaml-cpp/dll.h). It will also be necessary to correct the CMakeLists.txt, the solution can be viewed [here](https://github.com/hkarel/yaml-cpp/commit/b0771c76840a1f18303d11539441b7fe853b66cf)

**Comment by jbeder** [comment]
@hkarel @pinotree @PhilipDeegan If any of you has time and a Windows machine, could you please help getting PR #1063 passing the Windows shared lib test?

**Comment by jbeder** [comment]
Fixed by 2b65c65e1be6f2eec7a1ad6a91bc5a8765a7e4bd. Thanks @hkarel and @PhilipDeegan!

And once more to @pinotree: I'm open to a PR fixing the original issue you were trying to fix!
